### DISCUSSION TITLE
Working with the annotation of the PDF file using Page.first_annot

### DISCUSSION THREAD

**Discussion by mohitrobo** [pymupdf/PyMuPDF#1841] [Looking for help] [answered]
I was working on a PDF page as shown.
![Annotation](https://user-images.githubusercontent.com/37827176/181430768-3c07e1d0-9e38-4790-842b-c6335e491d73.jpg)

1.) The **underline annotation was created first**, while the highlight annotation was created afterward.
2.) When I use **"Page.first_annot"**, the first annotation that it returns is the underlined text and then the highlighted text.
3.) Going through the image, the first annotation has to be the highlighted text.
Why is it so and how can highlighted text be called first?

**Comment by JorjMcKie** [comment, marked as answer]
Annotations of a PDF page are stored within its object definition as an array of xref numbers that looks like so: `/Annots [ 5 0 R 9 0 R]`. Here you can deduct, that annot with xref 9 was created **_after_** annot with xref 5 had been created.
New annotations are appended to this array. So the sequence is no indication for the items' positions on the page.

> Also note that this array contains every object that PDF counts as an annotation type: links, fields, other annotations. In (Py-) MuPDF, these categories appear as logically different subtypes and are yielded under different Python iterators.

The page properties `first_link`, `first_annot`, `first_widget` and the associated iterators walk through this array and yield the first / next item of it, that has the correct type.
No attempt is being made to, for example, sort the array according to whatever criteria.

But you could do this when needed: `annots = sorted(page.annots(), key=lambda a: (a.rect.y1, a.rect.x0))`.
This list contains the annotations in the order top-left to bottom-right.
In your example, the highlight annotation would appear as `annots[0]`.

**Reply by mohitrobo** [reply]
Thank you [JorjMcKie](https://github.com/JorjMcKie)
